### ISSUE TITLE
no-redeclare identifies false positives with switch-case statements

### ISSUE THREAD

**Issue by scrouthtv** [eslint/eslint#13998]
I think I found a bug in the `no-redeclare` rule when using switch-case structures.
Thanks for your time and help in advance.

**Tell us about your environment**
```bash
 ~ eslint --env-info
Environment Info:

Node version: v15.5.1
npm version: v6.14.11
Local ESLint version: Not found
Global ESLint version: v7.17.0 (Currently used)
```

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**
I think I am using `default`. I did not change any settings, and I'm not using ts, but I'm not sure.

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```bash
 ~ cat .eslintrc.yml
env:
  browser: true
  es2021: true
  node: true
extends: 'eslint:recommended'
parserOptions:
  ecmaVersion: 12
  sourceType: module
rules: {}
```

</details>

```js
console.log(hello_world(1));

function hello_world(v) {
	var output = "";
	switch (v) {
		case 0:
			var i;
			for (i = 0; i < 4; i++)
				output += "0";
			break;
		case 1:
			console.log(i);
			var i;
			for (i = 0; i < 2; i++)
				output += "e";
			break;
		case 2:
			var i;
			for (i = 0; i < 8; i++)
				output += ".";
			break;
		default:
			output = "???";
	}
	return output;
}
```
This is an example file for my issue.
If you call hello_world with 1, the output is 
```
undefined
ee
```
So one would think that `i` is not declared if the whole case block is not evaluated.

```bash
 ~ eslint test.js
```

**What did you expect to happen?**
I'd expect this code to be completely valid and don't give any errors.
Especially shouldn't there bo any `redeclare` errors as no variables are redeclared

**What actually happened? Please include the actual, raw output from ESLint.**
```

/home/x/git/gotest/test.js
  13:8  error  'i' is already defined  no-redeclare
  18:8  error  'i' is already defined  no-redeclare

✖ 2 problems (2 errors, 0 warnings)

```

**Are you willing to submit a pull request to fix this bug?**
No. I don't know enough about js.

**Comment by anikethsaha** [comment]
Thanks for the report. 

IMO
This is working as expected. in your example, in same block (switch block) `I` is being declared twice. 
You can convert the case statements to block and then replace `var` with `let`.

[demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tcmVkZWNsYXJlOiBcImVycm9yXCIqL1xuXG5cbmZ1bmN0aW9uIGhlbGxvX3dvcmxkKHYpIHtcblx0dmFyIG91dHB1dCA9IFwiXCI7XG5cdHN3aXRjaCAodikge1xuXHRcdGNhc2UgMDoge1xuXHRcdFx0bGV0IGk7XG5cdFx0XHRmb3IgKGkgPSAwOyBpIDwgNDsgaSsrKVxuXHRcdFx0XHRvdXRwdXQgKz0gXCIwXCI7XG5cdFx0XHRicmVhaztcbiAgICAgICAgfVxuXHRcdGNhc2UgMToge1xuXHRcdFx0Y29uc29sZS5sb2coaSk7XG5cdFx0XHRsZXQgaTtcblx0XHRcdGZvciAoaSA9IDA7IGkgPCAyOyBpKyspXG5cdFx0XHRcdG91dHB1dCArPSBcImVcIjtcblx0XHRcdGJyZWFrO1xuICAgICAgICB9XG5cdFx0Y2FzZSAyOiB7XG5cdFx0XHRsZXQgaTtcblx0XHRcdGZvciAoaSA9IDA7IGkgPCA4OyBpKyspXG5cdFx0XHRcdG91dHB1dCArPSBcIi5cIjtcblx0XHRcdGJyZWFrO1xuICAgICAgICB9XG5cdFx0ZGVmYXVsdDpcblx0XHRcdG91dHB1dCA9IFwiPz8/XCI7XG5cdH1cblx0cmV0dXJuIG91dHB1dDtcbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjEyLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)

**Comment by scrouthtv** [comment]
Thanks for your answer.
I don't know much about js so I might be in the wrong here, but with the example I provided (switch-case **without** blocks), the `i` does not appear to be declared in the other case statement.
See https://jsfiddle.net/w6d0r9qh/, `i` gives undefined when printed before being declared in the same block (l. 12).

Since I have a `break` at the end of every `case`, I'm thinking of the whole construct to be equal to an `if`-`else if` construct.
So if the function is called with `v` set to 1, the `case 0:` isn't even run and as such also the `var i;` in l. 9 never evaluated.

Here's another example:
According to https://stackoverflow.com/a/519157/7242251
https://jsfiddle.net/w6d0r9qh/1/
a variable is defined (declared), if `typeof` doesn't return `undefined`.
In this example, `typeof` does return `undefined`, so I'd think that `i` does not exist in another block

**Comment by mdjermanovic** [comment]
`typeof` returns `undefined` for undeclared variables, too.

`var` variables declared in a function are function-scoped, so the rule is correct in this case: variable `i` is declared multiple times.

For example, the following would throw "ReferenceError: i is not defined" if we remove `var i`:

```js
function foo(v) {
    switch (v) {
        case 0:
            var i;
            break;
        case 1:
            i;
            break;
    }
}

foo(1);
```

**Comment by scrouthtv** [comment]
Thanks a lot for all of your help.
I've changed the switch-case to use blocks as suggested earlier.
I think that concludes my problem.
